slither/utils/upgradeability.py

```diff
@@ -0,0 +1,463 @@
+    return None
+
+
+def get_proxy_implementation_var(proxy: Contract) -> Optional[Variable]:
+    """
+    Gets the Variable that stores a proxy's implementation address. Uses data dependency to trace any LocalVariable
+    that is passed into a delegatecall as the target address back to its data source, ideally a StateVariable.
+    Args:
+        proxy: A Contract object (proxy.is_upgradeable_proxy should be true).
+
+    Returns:
+        (`Variable`) | None : The variable, ideally a StateVariable, which stores the proxy's implementation address.
+    """
+    available_functions = proxy.available_functions_as_dict()
+    if not proxy.is_upgradeable_proxy or not available_functions["fallback()"]:
+        return None
+
+    delegate = find_delegate_in_fallback(proxy)
+    if isinstance(delegate, LocalVariable):
+        dependencies = get_dependencies(delegate, proxy)
+        try:
+            delegate = next(var for var in dependencies if isinstance(var, StateVariable))
+        except StopIteration:
+            return delegate
+    return delegate
+
+
+def find_delegate_in_fallback(proxy: Contract) -> Optional[Variable]:
+    """
+    Searches a proxy's fallback function for a delegatecall, then extracts the Variable being passed in as the target.
+    Should typically be called by get_proxy_implementation_var(proxy).
+    Args:
+        proxy: A Contract object (should have a fallback function).
+
+    Returns:
+        (`Variable`) | None : The variable being passed as the destination argument in a delegatecall in the fallback.
+    """
+    delegate: Optional[Variable] = None
+    fallback = proxy.available_functions_as_dict()["fallback()"]
+    for node in fallback.all_nodes():
+        for ir in node.irs:
+            if isinstance(ir, LowLevelCall) and ir.function_name == "delegatecall":
+                delegate = ir.destination
+        if delegate is not None:
+            break
+        if (
+            node.type == NodeType.ASSEMBLY
+            and isinstance(node.inline_asm, str)
+            and "delegatecall" in node.inline_asm
+        ):
+            delegate = extract_delegate_from_asm(proxy, node)
+        elif node.type == NodeType.EXPRESSION:
+            expression = node.expression
+            if isinstance(expression, AssignmentOperation):
+                expression = expression.expression_right
+            if (
+                isinstance(expression, CallExpression)
+                and "delegatecall" in str(expression.called)
+                and len(expression.arguments) > 1
+            ):
+                dest = expression.arguments[1]
+                if isinstance(dest, CallExpression) and "sload" in str(dest.called):
+                    dest = dest.arguments[0]
+                if isinstance(dest, Identifier):
+                    delegate = dest.value
+                    break
+                if isinstance(dest, Literal) and len(dest.value) == 66:
+                    delegate = StateVariable()
```

Comment:
That's an interesting case that we dont usually use: having a StateVariable created on the fly by an analysis.

This might break assumption in other part of the codebase, we should probably document that `find_delegate_in_fallback` can return a newly created variable